File: 0020_valid_parentheses.py
# https://leetcode-cn.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s):
        """
        Inputs:
            s: str
        Outputs:
            res: bool
        """

        res = True
        m = {'}':'{', ')':'(', ']':'['}
        stack = []

        for i in s:
            if i not in m:
                stack.append(i)
            else:
                tmp = '-'
                if stack:
                    tmp = stack.pop()
                if tmp != m[i]:
                    res = False
                    break
        if stack:
            res = False
        return res

File: Cpp/0020_valid_parentheses.cpp
// https://leetcode.cn/problems/valid-parentheses/

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> table{
            {'}', '{'}, {']', '['}, {')', '('}
        };
        stack<char> stk;

        for (char c: s) {
            auto iter = table.find(c);
            if (iter != table.end()) {
                if ((stk.empty()) || (stk.top() != table[c])) {
                    return false;
                }

                stk.pop();
            }
            else {
                stk.push(c);
            }
        }

        return stk.empty();
    }
};